crates/accelerate/src/sabre/route.rs

```diff
@@ -0,0 +1,584 @@
+                                self.extended_set
+                                    .push([a.to_phys(&self.layout), b.to_phys(&self.layout)]);
+                                to_visit.push(successor_node);
+                                continue;
+                            }
+                        }
+                        visit_now.push(successor_node);
+                    }
+                }
+                j += 1;
+            }
+            visit_now.clear();
+            i += 1;
+        }
+        for (node, amount) in decremented.iter() {
+            self.required_predecessors[*node] += *amount;
+        }
+    }
+
+    /// Add swaps to the current set that greedily bring the nearest node together.  This is a
+    /// "release valve" mechanism; it ignores all the Sabre heuristics and forces progress, so we
+    /// can't get permanently stuck.
+    fn force_enable_closest_node(
+        &mut self,
+        current_swaps: &mut Vec<[PhysicalQubit; 2]>,
+    ) -> NodeIndex {
+        let (&closest_node, &qubits) = {
+            let dist = &self.target.distance;
+            self.front_layer
+                .iter()
+                .min_by(|(_, qubits_a), (_, qubits_b)| {
+                    dist[[qubits_a[0].index(), qubits_a[1].index()]]
+                        .partial_cmp(&dist[[qubits_b[0].index(), qubits_b[1].index()]])
+                        .unwrap_or(Ordering::Equal)
+                })
+                .unwrap()
+        };
+        let shortest_path = {
+            let mut shortest_paths: DictMap<NodeIndex, Vec<NodeIndex>> = DictMap::new();
+            (dijkstra(
+                &self.target.coupling,
+                NodeIndex::new(qubits[0].index()),
+                Some(NodeIndex::new(qubits[1].index())),
+                |_| Ok(1.),
+                Some(&mut shortest_paths),
+            ) as PyResult<Vec<Option<f64>>>)
+                .unwrap();
+            shortest_paths
+                .get(&NodeIndex::new(qubits[1].index()))
+                .unwrap()
+                .iter()
+                .map(|n| PhysicalQubit::new(n.index() as u32))
+                .collect::<Vec<_>>()
+        };
+        // Insert greedy swaps along that shortest path, splitting them between moving the left side
+        // and moving the right side to minimise the depth.  One side needs to move up to the split
+        // point and the other can stop one short because the gate will be routable then.
+        let split: usize = shortest_path.len() / 2;
+        current_swaps.reserve(shortest_path.len() - 2);
+        for i in 0..split {
+            current_swaps.push([shortest_path[i], shortest_path[i + 1]]);
+        }
+        for i in 0..split - 1 {
+            let end = shortest_path.len() - 1 - i;
+            current_swaps.push([shortest_path[end], shortest_path[end - 1]]);
+        }
+        current_swaps.iter().for_each(|&swap| self.apply_swap(swap));
+        closest_node
+    }
```

Comment:
Previously, this release valve was three separate bits of logic in the main logic loop - `closest_operation`, `swaps_to_route` and then the application of those swaps.  Those three were all super linked, so I pulled them into one single function.